colorbleed/plugins/maya/publish/validate_animation_out_set_related_node_ids.py

```diff
@@ -3,87 +3,127 @@
 
-    Note that this doesn't ensure it's from the current Task. An ID created
-    from `lookdev` has the same relation to the Item as one coming from others,
-    like `rigging` or `modeling`.
+class ValidateAnimationOutSetRelatedNodeIds(pyblish.api.InstancePlugin):
+    """Validate if nodes have related IDs to the source (original shapes)
+
+    Any intermediate shapes which are created when creating deformers on
+    shapes will need to get the correct ID to ensure the look assignment still
+    works on the new shape.
 
     """
 
     order = colorbleed.api.ValidateContentsOrder
-    families = ['colorbleed.animation']
+    families = ['colorbleed.animation', "colorbleed.pointcache"]
     hosts = ['maya']
     label = 'Animation Out Set Related Node Ids'
-    actions = [colorbleed.api.SelectInvalidAction]
+    actions = [colorbleed.api.SelectInvalidAction, colorbleed.api.RepairAction]
     optional = True
 
     ignore_types = ("constraint",)
 
     def process(self, instance):
         """Process all meshes"""
 
-        # Ensure all nodes have a cbId
+        # Ensure all nodes have a cbId and a related ID to the original shapes
+        # if a deformer has been created on the shape
         invalid = self.get_invalid(instance)
         if invalid:
             raise RuntimeError("Nodes found with non-related "
                                "asset IDs: {0}".format(invalid))
 
     @classmethod
-    def get_pointcache_nodes(cls, instance):
-
-        # Get out_SET
-        sets = cmds.ls(instance, type='objectSet')
-        pointcache_sets = [x for x in sets if x == 'out_SET']
-
-        nodes = list()
-        for s in pointcache_sets:
-            members = cmds.sets(s, query=True)
-            members = cmds.ls(members, long=True)  # ensure long names
-            descendents = cmds.listRelatives(members,
-                                             allDescendents=True,
-                                             fullPath=True) or []
-            descendents = cmds.ls(descendents,
-                                  noIntermediate=True,
-                                  long=True)
-            hierarchy = members + descendents
-            nodes.extend(hierarchy)
-
-        # ignore certain node types (e.g. constraints)
-        ignore = cmds.ls(nodes, type=cls.ignore_types, long=True)
-        if ignore:
-            ignore = set(ignore)
-            nodes = [node for node in nodes if node not in ignore]
-
-        return nodes
-
-    @classmethod
     def get_invalid(cls, instance):
-        invalid_items = []
+        """Get all nodes which do not match the criteria"""
+        invalid = []
 
         # get asset id
-        nodes = cls.get_pointcache_nodes(instance)
+        nodes = instance.data.get("out_hierarchy", instance[:])
         for node in nodes:
+            node_type = cmds.nodeType(node)
             node_id = lib.get_id(node)
+
+            if node_type == "mesh" and not node_id:
```

Comment:
This is explicit, but I'm not sure it's safe enough. Aren't there more types than just meshes that require ids? Or is this full validator only *ever* checking meshes? - If it's meshes only should the class and docstring maybe match that.

Looking at the class and docstring I believe it's a general safety net nodes having their id?
Looking at the code it seems to be more a safety net for "deformed meshes" missing their id.

I'm not sure what the intentions are anymore. :)

Reply:
I have made it less explicit but still pretty explicit with `obj_type = cmds.objectType(node, isAType="shape")` we want to check the shapes because the shapes get the shader ;)

Reply:
> I have made it less explicit but still pretty explicit with obj_type = cmds.objectType(node, isAType="shape") we want to check the shapes because the shapes get the shader ;)

Sure, I'm fine with explicit. The problem is more that the class/docstring wasn't described as one that is *so explicit*. With this type check the validation is *not* generic at all, it only considers shapes. If we go down that route I'd say it even needs to be "deformable shapes" and the validation is intended to only check for "lost ids" because of deformers.

Reply:
Also note that a parent constraint is also a "shape", I think deformers are to. Maybe this type check now is to broad for Maya to be accurate?